File: server/src/main/java/com/mxlm/card/PlaywrightPool.java
package com.mxlm.card;

import com.microsoft.playwright.*;
import com.microsoft.playwright.options.LoadState;
import com.microsoft.playwright.options.ScreenshotType;
import jakarta.annotation.PostConstruct;
import jakarta.annotation.PreDestroy;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

/**
 * Playwright 浏览器池（单例）
 * <p>
 * Playwright 启动成本高（首次要下载 Chromium），故用单例保持一个 Browser 实例。
 * 每次截图创建独立 BrowserContext，保证隔离。
 * <p>
 * 若首次运行未安装浏览器，会自动执行 `playwright install chromium`。
 */
@Slf4j
@Component
public class PlaywrightPool {

    @Value("${card.playwright-enabled:true}")
    private boolean enabled;

    private Playwright playwright;
    private Browser browser;

    @PostConstruct
    public void init() {
        if (!enabled) {
            log.warn("[PlaywrightPool] Playwright 已禁用，卡片生成功能不可用");
            return;
        }
        try {
            log.info("[PlaywrightPool] 初始化 Playwright...");
            playwright = Playwright.create();
            browser = tryLaunchChromium();
            log.info("[PlaywrightPool] Playwright 初始化成功");
        } catch (Exception e) {
            // 首次运行 / 构建期未预下载：尝试运行时自动下载 Chromium
            log.warn("[PlaywrightPool] 首次启动检测到 Chromium 未安装，尝试自动下载... ({})", e.getMessage());
            try {
                // Playwright Java CLI 会下载 Chromium 到 PLAYWRIGHT_BROWSERS_PATH（默认 ~/.cache/ms-playwright）
                com.microsoft.playwright.CLI.main(new String[]{"install", "chromium"});
                log.info("[PlaywrightPool] Chromium 自动下载完成，重新初始化...");
                if (playwright == null) {
                    playwright = Playwright.create();
                }
                browser = tryLaunchChromium();
                log.info("[PlaywrightPool] Playwright 初始化成功（运行时下载模式）");
            } catch (Throwable retryEx) {
                log.error("[PlaywrightPool] Playwright 初始化失败，卡片生成将不可用。" +
                        "请检查网络或手动执行: mvn exec:java -D exec.mainClass=com.microsoft.playwright.CLI -D exec.args=\"install chromium\"", retryEx);
            }
        }
    }

    /**
     * 启动 Chromium 浏览器
     */
    private Browser tryLaunchChromium() {
        return playwright.chromium().launch(new BrowserType.LaunchOptions()
                .setHeadless(true)
                .setArgs(java.util.List.of(
                        "--no-sandbox",
                        "--disable-dev-shm-usage",
                        "--disable-gpu"
                )));
    }

    /**
     * 将 HTML 渲染成 PNG 字节数组
     *
     * @param html   完整 HTML 内容
     * @param width  视口宽度
     * @param height 视口高度
     * @return PNG 字节数据
     */
    public byte[] renderHtmlToPng(String html, int width, int height) {
        if (browser == null) {
            throw new IllegalStateException("Playwright 未初始化");
        }
        try (BrowserContext context = browser.newContext(new Browser.NewContextOptions()
                .setViewportSize(width, height)
                .setDeviceScaleFactor(2.0)); // 2 倍图更清晰
             Page page = context.newPage()) {

            page.setContent(html);
            // 等 fonts + 图片加载完成
            page.waitForLoadState(LoadState.NETWORKIDLE);
            // 额外等 200ms 确保字体渲染稳定
            page.waitForTimeout(200);

            return page.screenshot(new Page.ScreenshotOptions()
                    .setType(ScreenshotType.PNG)
                    .setFullPage(false)
                    .setOmitBackground(false));
        }
    }

    @PreDestroy
    public void destroy() {
        try {
            if (browser != null) {
                browser.close();
            }
            if (playwright != null) {
                playwright.close();
            }
            log.info("[PlaywrightPool] 已关闭");
        } catch (Exception e) {
            log.warn("[PlaywrightPool] 关闭异常", e);
        }
    }
}

File: server-py/app/playwright_pool.py
"""
Playwright 无头浏览器单例池
对应 Java 的 PlaywrightPool

关键设计：
- FastAPI 运行在 asyncio 事件循环里，sync_playwright() 无法直接在该循环里 start
- 解决方案：把 Playwright 的所有同步调用都派发到一个独立的专用线程里执行
  该线程没有 asyncio 事件循环，因此 sync_playwright 可以正常工作
- 用队列 (concurrent.futures) 让外部（异步/同步）都能安全地调用截图
"""
from __future__ import annotations

import logging
import threading
from concurrent.futures import ThreadPoolExecutor, Future
from typing import Optional

from playwright.sync_api import Browser, Playwright, sync_playwright

from app.config import settings

logger = logging.getLogger(__name__)


class PlaywrightPool:
    """浏览器单例池（所有 Playwright 操作都在专用单线程内执行）"""

    def __init__(self) -> None:
        self._playwright: Optional[Playwright] = None
        self._browser: Optional[Browser] = None
        self._render_lock = threading.Lock()
        self._enabled = settings.card_playwright_enabled
        # 关键：单线程 Executor —— 保证所有 Playwright 调用都在同一个非 asyncio 线程里
        self._executor: Optional[ThreadPoolExecutor] = None
        self._started = False

    # ---------- 内部：真正在专用线程里执行的方法 ----------
    def _do_start(self) -> None:
        logger.info("[PlaywrightPool] 初始化 Playwright...")
        self._playwright = sync_playwright().start()
        self._browser = self._playwright.chromium.launch(
            headless=True,
            args=[
                "--no-sandbox",
                "--disable-dev-shm-usage",
                "--disable-gpu",
            ],
        )
        logger.info("[PlaywrightPool] Playwright 初始化成功")

    def _do_stop(self) -> None:
        try:
            if self._browser:
                self._browser.close()
                self._browser = None
            if self._playwright:
                self._playwright.stop()
                self._playwright = None
            logger.info("[PlaywrightPool] 已关闭")
        except Exception as e:
            logger.warning("[PlaywrightPool] 关闭异常：%s", e)

    def _do_render(self, html: str, width: int, height: int) -> bytes:
        if self._browser is None:
            raise RuntimeError("Playwright 未初始化")
        context = self._browser.new_context(
            viewport={"width": width, "height": height},
            device_scale_factor=2.0,
        )
        try:
            page = context.new_page()
            page.set_content(html, wait_until="networkidle")
            page.wait_for_timeout(200)
            return page.screenshot(type="png", full_page=False, omit_background=False)
        finally:
            context.close()

    # ---------- 对外：把调用派发到专用线程 ----------
    def start(self) -> None:
        """启动 Playwright（应用启动时调用）—— 惰性初始化，失败不影响服务启动"""
        if not self._enabled:
            logger.warning("[PlaywrightPool] Playwright 已禁用，卡片功能不可用")
            return
        if self._started:
            return
        try:
            self._executor = ThreadPoolExecutor(
                max_workers=1,
                thread_name_prefix="playwright-worker",
            )
            fut: Future = self._executor.submit(self._do_start)
            fut.result(timeout=60)  # 等待启动完成，超时抛异常
            self._started = True
        except Exception as e:
            logger.exception("[PlaywrightPool] 初始化失败：%s", e)
            logger.error("首次运行请执行: playwright install chromium")
            # 保证探针能通：不抛出，卡片功能降级
            if self._executor is not None:
                self._executor.shutdown(wait=False, cancel_futures=True)
                self._executor = None

    def stop(self) -> None:
        """关闭浏览器（应用停机时调用）"""
        if self._executor is None:
            return
        try:
            fut = self._executor.submit(self._do_stop)
            fut.result(timeout=15)
        except Exception as e:
            logger.warning("[PlaywrightPool] 关闭异常：%s", e)
        finally:
            self._executor.shutdown(wait=False, cancel_futures=True)
            self._executor = None
            self._started = False

    def render_html_to_png(self, html: str, width: int, height: int) -> bytes:
        """将 HTML 渲染为 PNG 字节数组（线程安全，串行执行）"""
        if not self._started or self._executor is None:
            raise RuntimeError(
                "Playwright 未初始化成功，卡片功能不可用；请检查容器内是否安装了 chromium"
            )
        # 加锁保证同一时刻只有一个截图任务在专用线程里跑
        with self._render_lock:
            fut: Future = self._executor.submit(self._do_render, html, width, height)
            return fut.result(timeout=60)


pool = PlaywrightPool()
